cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -199,253 +214,249 @@ __global__ void subtract_set_bits_range_boundaries_kernel(bitmask_type const* bi
 {
   constexpr size_type const word_size_in_bits{detail::size_in_bits<bitmask_type>()};
 
-  cudf::size_type const tid = threadIdx.x + blockIdx.x * blockDim.x;
-  cudf::size_type range_id  = tid;
+  size_type const tid = threadIdx.x + blockIdx.x * blockDim.x;
+  size_type range_id  = tid;
 
   while (range_id < num_ranges) {
     size_type const first_bit_index = *(first_bit_indices + range_id);
```

Comment:
I thought about this too. In my understanding, this function takes an arbitrary iterator, which may or may not behave the same, since `first_bit_indices[range_id]` could go through the `operator[]` of whatever type of iterator is passed in, rather than incrementing the iterator and dereferencing the result. In some cases I would expect these to behave identically, but I think the existing syntax is important to make it work for arbitrary iterator types.

Reply:
Following up - I spent some time trying to find conclusive information about what kinds of iterators should support `iterator[index]`. I could not find an authoritative reference but I built some code samples to test this question. It seems like boost's counting iterator supports the indexing notation (https://www.godbolt.org/z/7TP1E5cdb) but arbitrary iterators do not support the notation `iterator[index]` just from defining the dereference and increment operators (https://godbolt.org/z/T8GY8f7bs). I think this is a good argument for keeping this as-is.

Reply:
In terms of standard iterator types, the only type for which `*(iterator + index)` is valid is a `RandomAccessIterator`, which is the same for `iterator[index]`. Even `BidirectionalIterators` are only guaranteed to offer the (pre-/post-) increment and decrement operators (`ForwardIterators` do not support decrementing). The documentation here claims that `first_bit_indices` is a random access iterator, so if we intend to also support simpler concepts then this line should probably use something like [`thrust::advance`](https://thrust.github.io/doc/group__iterators_gafd488bbe514cc2cc3ff28401c0229685.html#gafd488bbe514cc2cc3ff28401c0229685), which will work for any `InputIterator` (at the expense of some performance).

Reply:
Random access iterators require `operator[]`. Other iterators do not.

Reply:
@jrhemstad you mean `operator++`, not `operator+`, right? And `thrust::advance` is guaranteed to work for `InputIterators` (which internally I assume is equivalent to `for (i = 0; i < n; ++i) it++;`), so that would be an idiomatic choice if we really need this kind of offset here. From a cursory inspection of the algorithm it seems like the way that work is distributed across threads requires skipping forward at the end of each iteration of the while loop, so that would be a suitable place to call `thrust::advance`.

@bdice requested some references offline, here's [a definition of a random access iterator](https://en.cppreference.com/w/cpp/iterator/random_access_iterator) (others can be found in the same standard header). This reference shows that random access iterators are the only ones providing `operator[]` and `operator+`, which therefore cannot be relied on here.

Reply:
To be clear, pretty much every Thrust/libcudf algorithm operating on iterators requires random access iterators, so it's not really onerous to have the same expectation here. I main goal was to convey that using `*(it+i)` is more canonical for iterator algorithms than `it[i]`.